content/languages/c/authoring/index.md

```diff
@@ -0,0 +1,285 @@
+- Some header files include other header files indirectly, for example, file `foo.h` contains line `#include <bar.h>`, which might appear to make the explicit include for `bar.h` unnecessary. It's not true though, because the file `foo.h` might change one day, or might depend on some compiler settings or command line options, and after some changes to the configuration of the C runner, the `bar.h` might be not included there anymore. That's why every file (i.e. code snippet) of a kata should explicitly include all required header files declaring functions used in it.
+- The author might think that header files for the testing framework are included automatically by the code runner. That is not the case though, and test suites need to include `criterion/criterion.h` explicitly.
+
+### Compilation warnings
+
+Compiler options related to warnings used by the C runner are somewhat strict and require some discipline to get the code to compile cleanly. `-Wall` and `-Wextra` may cause numerous warnings and some of them are very pedantic. However, code of C kata should still compile cleanly, without any warnings logged to the console. Even when a warning does not cause any problem with tests, users get distracted by them and blame them for failing tests.
+
+
+### Memory management
+
+Unlike many modern, high-level languages, C does not manage memory automatically. Manual memory management is a very vast and complex topic, with many possible ways of achieving the goal depending on a specific case, caveats, and pitfalls.
+
+Whenever a kata needs to return a string or an array, C authors tend to use the naive technique of allocating the memory in the solution function, and freeing it in the test suite. This approach mimics the behavior known from other languages where returning an array or object from inside of the user's solution is perfectly valid, but it's hardly ever a valid way of working with unmanaged memory.
+
+One of the consequences of unmanaged memory is that it's strongly recommended against returning string constants from C functions, especially when translating kata from other languages. Returning a string in other languages is not a problem, but in C it always raises questions of who should allocate it and how it should be allocated. Consider replacing the string with some simpler data type (eventually aliased with a `typedef`), and/or provide some symbolic constants for available values. For example, if the requirement for the Python version is to _Return the string `'You won!'` if you won the game, `'You lost :('` if you were defeated, and `'It's a draw'` if there is no winner._, C should preferably use the constants `WON`, `LOST` and `DRAW`. If the author decides to keep raw C-strings as elements of the kata interface, they should clearly specify the required allocation scheme.
```

Comment:
Is there a reason why plain constants are preferred over `enum`s for a user-defined set of values (referring to "C should preferably use the constants `WON`, `LOST` and `DRAW`")? IMO the latter better communicates the intent that these user-defined values are related to each other and constitute a "type" of their own.

Reply:
Have you seen my comment [here](https://github.com/codewars/docs/pull/228#discussion_r554881117)?

Of course, from the design point of view, you are completely right. However, I am somewhat concerned with how such guideline would be received by Codewars authors. I can imagine a situation when a Python kata, for example [Red Knight](https://www.codewars.com/kata/5fc4349ddb878a0017838d0f/), has a return type of `(string, int)` and now C translator would like to have return values of `{ "White", 10 }` or `{"Black", 42}` (i.e. to keep the requirement of having a string, what was actually the case), instead of introducing an enum. I do not know why they would want to, but I've already seen various explanations, starting from "because consistency" to "`char` abbreviation will be OK I like it more`" to "I don't like enums". My goal for this particular guideline is to remove the ambuiguity of the allocation type which should be used, and not enforce some particular coding style. As much as I personally would prefer, use, and suggest an enum here, I do not want to strictly force authors to do so and I want to give them a choice.

As I said, I do not see any problem with putting your part on enums here other than authors feeling forced. You can re-suggest it, and I will merge it.

Reply:
Point taken - let's leave this section as is :+1: